### PR TITLE
feat(connector):  add support for bank redirect for Paypal

### PR DESCRIPTION
**Pull request by prasunna09**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Add bank redirects - Ideal, Eps, Giropay, and Sofort for Paypal connector.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Tested using postman.
Screenshot 2023-05-10 at 2 54 12 PM

Screenshot 2023-05-10 at 2 55 11 PM

Tested a successful payment
Screenshot 2023-05-10 at 2 55 52 PM

Screenshot 2023-05-10 at 2 56 26 PM

Tested a failed payment
Screenshot 2023-05-10 at 2 59 02 PM

Screenshot 2023-05-10 at 2 59 23 PM

UI Tests for Bank Redirects
Screen Shot 2023-07-28 at 4 18 19 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/api_models/src/payments.rs [hunk 1]
```diff
@@ -566,15 +564,9 @@ impl From<&PaymentMethodData> for AdditionalPaymentData {
...
-            PaymentMethodData::BankRedirect(bank_redirect_data) => match bank_redirect_data {
-                BankRedirectData::Eps { bank_name, .. } => Self::BankRedirect {
-                    bank_name: Some(bank_name.to_owned()),
-                },
-                BankRedirectData::Ideal { bank_name, .. } => Self::BankRedirect {
-                    bank_name: Some(bank_name.to_owned()),
-                },
-                _ => Self::BankRedirect { bank_name: None },
-            },
+            PaymentMethodData::BankRedirect(bank_redirect_data) => {
```
Comment (jagan-jaya) [inline crates/api_models/src/payments.rs:567]: is this discussed with @Narayanbhat166 , will it break any other flow?

File: crates/router/src/compatibility/stripe/errors.rs [hunk 2]
```diff
@@ -82,6 +82,9 @@ pub enum StripeErrorCode {
     #[error(error_type = StripeErrorType::InvalidRequestError, code = "resource_missing", message = "No such payment method")]
     PaymentMethodNotFound,
 
+    #[error(error_type = StripeErrorType::InvalidRequestError, code = "resource_missing", message = "No such payment method")]
```
Comment (jagan-jaya) [inline crates/router/src/compatibility/stripe/errors.rs:85]:
```suggestion
    #[error(error_type = StripeErrorType::InvalidRequestError, code = "resource_missing", message = "No such payment method type")]
```

File: crates/router/src/connector/paypal/transformers.rs [hunk 3]
```diff
@@ -93,6 +99,58 @@ fn get_address_info(
...
+        BankRedirectData::Sofort {
+            country,
+            preferred_language: _,
+            billing_details,
+        } => Ok(PaymentSourceItem::Sofort(RedirectRequest {
+            name: billing_details.billing_name.clone(),
+            country_code: *country,
+            experience_context: ContextStruct {
+                return_url: item.request.complete_authorize_url.clone(),
+                cancel_url: item.request.complete_authorize_url.clone(),
```
Comment (jagan-jaya) [inline crates/router/src/connector/paypal/transformers.rs:148]: is cancel flow tested?

File: crates/router/src/core/payments/helpers.rs [hunk 4]
```diff
@@ -1252,7 +1252,8 @@ pub(crate) fn validate_pm_or_token_given(
             Some(api_enums::PaymentMethodType::Paypal)
         ) && !matches!(mandate_type, Some(api::MandateTxnType::RecurringMandateTxn))
             && token.is_none()
-            && (payment_method_data.is_none() || payment_method.is_none()),
+            && (payment_method_data.is_none() || payment_method.is_none())
+            && !matches!(payment_method, Some(api_enums::PaymentMethod::BankRedirect)),
```
Comment (jagan-jaya) [inline crates/router/src/core/payments/helpers.rs:1256]: why there is a change here? aren't you getting payment_method_data and payment_method

File: crates/router/src/core/payments/operations/payment_complete_authorize.rs [hunk 5]
```diff
@@ -100,9 +100,20 @@ impl<F: Send + Clone> GetTracker<F, PaymentData<F>, api::PaymentsRequest> for Co
         let token = token.or_else(|| payment_attempt.payment_token.clone());
 
         helpers::validate_pm_or_token_given(
-            &request.payment_method,
+            &request.payment_method.or(Some(
+                payment_attempt
+                    .payment_method
+                    .ok_or(errors::ApiErrorResponse::PaymentMethodNotFound)?
+                    .foreign_into(),
```
Comment (jagan-jaya) [inline crates/router/src/core/payments/operations/payment_complete_authorize.rs:107]: use foreign_from

File: crates/router/src/core/payments/operations/payment_complete_authorize.rs [hunk 6]
```diff
@@ -100,9 +100,20 @@ impl<F: Send + Clone> GetTracker<F, PaymentData<F>, api::PaymentsRequest> for Co
...
+            )),
             &request.payment_method_data,
-            &request.payment_method_type,
+            &request.payment_method_type.or(Some(
+                payment_attempt
+                    .payment_method_type
+                    .clone()
+                    .ok_or(errors::ApiErrorResponse::PaymentMethodTypeNotFound)?
+                    .foreign_into(),
```
Comment (jagan-jaya) [inline crates/router/src/core/payments/operations/payment_complete_authorize.rs:115]: use foreign_from

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/paypal/transformers.rs [hunk 7]
```diff
@@ -93,6 +99,58 @@ fn get_address_info(
...
+            },
+        })),
+        BankRedirectData::Giropay {
+            billing_details, ..
+        } => Ok(PaymentSourceItem::Giropay(RedirectRequest {
+            name: billing_details.billing_name.clone(),
+            country_code: api_models::enums::CountryAlpha2::DE,
+            experience_context: ContextStruct {
+                return_url: item.request.complete_authorize_url.clone(),
+                cancel_url: item.request.complete_authorize_url.clone(),
```
Comment (jagan-jaya) [inline crates/router/src/connector/paypal/transformers.rs:125]: if cancellation is possible then it it handled in complete_authorize_flow?

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/paypal/transformers.rs (hunk 3 above)
Comment (prasunna09) [reply crates/router/src/connector/paypal/transformers.rs:148]: yes!

**Review by jagan-jaya — DISMISSED** [review]
Other than cancellation feature handling in complete_authorize, it looks good to me

**Review by ArjunKarthik — DISMISSED** [review]
Please confirm your changes on new stripe error with abhishek

File: crates/router/src/compatibility/stripe/errors.rs [hunk 8]
```diff
@@ -82,6 +82,9 @@ pub enum StripeErrorCode {
     #[error(error_type = StripeErrorType::InvalidRequestError, code = "resource_missing", message = "No such payment method")]
     PaymentMethodNotFound,
 
+    #[error(error_type = StripeErrorType::InvalidRequestError, code = "resource_missing", message = "No such payment method type")]
+    PaymentMethodTypeNotFound,
```
[the remark this answers is not in the document]
Comment (prasunna09) [reply crates/router/src/compatibility/stripe/errors.rs:86]: So I have removed this error in the other pr as it is no longer needed!

**Review by ArjunKarthik — DISMISSED** [review]

File: crates/api_models/src/payments.rs [hunk 9]
```diff
@@ -546,9 +546,7 @@ pub enum AdditionalPaymentData {
         card_issuer: Option<String>,
         card_network: Option<String>,
     },
-    BankRedirect {
-        bank_name: Option<api_enums::BankNames>,
-    },
+    BankRedirect(BankRedirectData),
```
Comment (jarnura) [inline crates/api_models/src/payments.rs:549]: Is this changes are discussed with dependent teams?

File: crates/router/src/connector/paypal/transformers.rs [hunk 10]
```diff
@@ -93,6 +99,74 @@ fn get_address_info(
...
+        BankRedirectData::Eps {
+            billing_details,
+            bank_name: _,
+        } => Ok(PaymentSourceItem::Eps(RedirectRequest {
+            name: billing_details.billing_name.clone().ok_or(
+                errors::ConnectorError::MissingRequiredField {
+                    field_name: "billing_name",
+                },
+            )?,
+            country_code: api_models::enums::CountryAlpha2::AT,
```
Comment (jarnura) [inline crates/router/src/connector/paypal/transformers.rs:116]: Why we are hard-coding country here?

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/paypal/transformers.rs [hunk 11]
```diff
@@ -159,6 +245,25 @@ impl TryFrom<&types::PaymentsAuthorizeRouterData> for PaypalPaymentsRequest {
                     "Payment Method".to_string(),
                 ))?,
             },
+            api::PaymentMethodData::BankRedirect(ref bank_redirection_data) => {
+                let intent = PaypalPaymentIntent::Capture;
```
Comment (jagan-jaya) [inline crates/router/src/connector/paypal/transformers.rs:249]: why are we hardcoding `capture`?

**Review by jagan-jaya — CHANGES_REQUESTED** [review]
@prasunna09 Add comments in the description on why vault changes are need here. 
Also add the testcases in [hs-tool](https://hs-payments-test.netlify.app)

**Review by jarnura — APPROVED** [review]

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]
Please resolve open comments

**Review by jarnura — APPROVED** [review]

**Review by jagan-jaya — APPROVED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Pull request merged**